--- a/acme.sh
+++ b/acme.sh
@@ -28,6 +28,8 @@
 
 DEFAULT_CA=$LETSENCRYPT_CA_V2
 DEFAULT_STAGING_CA=$LETSENCRYPT_STAGING_CA_V2
+
+DEFAULT_AGREEMENT="https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf"
 
 DEFAULT_USER_AGENT="$PROJECT_NAME/$VER ($PROJECT)"
 DEFAULT_ACCOUNT_EMAIL=""
@@ -1982,6 +1984,17 @@
       if [ "$url" = "$ACME_NEW_ACCOUNT" ] || [ "$url" = "$ACME_REVOKE_CERT" ]; then
         protected="$JWK_HEADERPLACE_PART1$nonce\", \"url\": \"${url}$JWK_HEADERPLACE_PART2, \"jwk\": $jwk"'}'
       else
+        if [ -z "${ACCOUNT_URL}" ]; then
+          _info "Re-reading ACCOUNT_URL"
+          _debug2 "ACCOUNT_URL was empty!"
+          _accUri=$(_readcaconf "ACCOUNT_URL")
+          export ACCOUNT_URL="${_accUri}"
+          _debug2 ACCOUNT_URL "${ACCOUNT_URL}"
+          if [ -z "${ACCOUNT_URL}" ]; then
+            _err "Cannot locate account URL."
+            return 1
+          fi
+        fi
         protected="$JWK_HEADERPLACE_PART1$nonce\", \"url\": \"${url}$JWK_HEADERPLACE_PART2, \"kid\": \"${ACCOUNT_URL}\""'}'
       fi
     else
@@ -2602,6 +2615,9 @@
   if [ -z "$domain" ]; then
     return 0
   fi
+
+  domain="$1"
+  _ilength="$2"
 
   if [ -z "$DOMAIN_PATH" ]; then
     domainhome="$CERT_HOME/$domain"
@@ -3156,30 +3172,56 @@
 # webroot  removelevel tokenfile
 _clearupwebbroot() {
   __webroot="$1"
+  __domain="$4"
   if [ -z "$__webroot" ]; then
     _debug "no webroot specified, skip"
     return 0
   fi
 
-  _rmpath=""
-  if [ "$2" = '1' ]; then
-    _rmpath="$__webroot/.well-known"
-  elif [ "$2" = '2' ]; then
-    _rmpath="$__webroot/.well-known/acme-challenge"
-  elif [ "$2" = '3' ]; then
-    _rmpath="$__webroot/.well-known/acme-challenge/$3"
-  else
-    _debug "Skip for removelevel:$2"
-  fi
-
-  if [ "$_rmpath" ]; then
-    if [ "$DEBUG" ]; then
-      _debug "Debugging, skip removing: $_rmpath"
-    else
-      rm -rf "$_rmpath"
-    fi
-  fi
-
+	if [ "$vtype" = "$VTYPE_HTTP" ]; then
+		h_api="$(_findHook "$d" httpapi "$_currentRoot")"
+		_debug h_api "$h_api"
+		if [ "$h_api" ]; then
+		  _info "Found domain http api file: $h_api"
+		  (
+			if ! . "$h_api"; then
+			  _err "Load file $h_api error. Please check your api file and try again."
+			  return 1
+			fi
+
+			rmcommand="${_currentRoot}_rm"
+			if ! _exists "$rmcommand"; then
+			  _err "It seems that your api file is not correct, it must have a function named: $rmcommand"
+			  return 1
+			fi
+
+			if ! $rmcommand "$__domain" "$3"; then
+			  _err "Error rm webroot api for domain:$__webroot"
+			  return 1
+			fi
+		  )
+		fi
+
+	else
+		_rmpath=""
+		if [ "$2" = '1' ]; then
+		  _rmpath="$__webroot/.well-known"
+		elif [ "$2" = '2' ]; then
+		  _rmpath="$__webroot/.well-known/acme-challenge"
+		elif [ "$2" = '3' ]; then
+		  _rmpath="$__webroot/.well-known/acme-challenge/$3"
+		else
+		  _debug "Skip for removelevel:$2"
+		fi
+
+		if [ "$_rmpath" ]; then
+		  if [ "$DEBUG" ]; then
+			_debug "Debugging, skip removing: $_rmpath"
+		  else
+			rm -rf "$_rmpath"
+		  fi
+		fi
+	fi
   return 0
 
 }
@@ -3850,7 +3892,7 @@
   _debug _main_domain "$_main_domain"
   _debug _alt_domains "$_alt_domains"
 
-  _key_length="$4"
+  _key_length="${4}"
   _real_cert="$5"
   _real_key="$6"
   _real_ca="$7"
@@ -3961,6 +4003,11 @@
   else
     _key=$(_readdomainconf Le_Keylength)
     _debug "Read key length:$_key"
+
+    if [ -z "${_key_length}" ] && [ ! -z "${_key}" ]; then
+      _key_length="${_key}"
+    fi
+
     if [ ! -f "$CERT_KEY_PATH" ] || [ "$_key_length" != "$_key" ] || [ "$Le_ForceNewDomainKey" = "1" ]; then
       if ! createDomainKey "$_main_domain" "$_key_length"; then
         _err "Create domain key error."
@@ -4329,7 +4376,30 @@
     _debug "_currentRoot" "$_currentRoot"
 
     if [ "$vtype" = "$VTYPE_HTTP" ]; then
-      if [ "$_currentRoot" = "$NO_VALUE" ]; then
+      h_api="$(_findHook "$d" httpapi "$_currentRoot")"
+      _debug h_api "$h_api"
+
+      if [ "$h_api" ]; then
+        _info "Found domain http api file: $h_api"
+        (
+          if ! . "$h_api"; then
+            _err "Load file $h_api error. Please check your api file and try again."
+            return 1
+          fi
+
+          addcommand="${_currentRoot}_add"
+          if ! _exists "$addcommand"; then
+            _err "It seems that your api file is not correct, it must have a function named: $addcommand"
+            return 1
+          fi
+
+          if ! $addcommand "$d" "$token" "$keyauthorization"; then
+            _err "Error add webroot for domain:$d"
+            return 1
+          fi
+        )
+
+      elif [ "$_currentRoot" = "$NO_VALUE" ]; then
         _info "Standalone mode server"
         _ncaddr="$(_getfield "$_local_addr" "$_ncIndex")"
         _ncIndex="$(_math $_ncIndex + 1)"
@@ -4384,7 +4454,7 @@
 
         if ! printf "%s" "$keyauthorization" >"$wellknown_path/$token"; then
           _err "$d:Can not write token to file : $wellknown_path/$token"
-          _clearupwebbroot "$_currentRoot" "$removelevel" "$token"
+          _clearupwebbroot "$_currentRoot" "$removelevel" "$token" "$d"
           _clearup
           _on_issue_err "$_post_hook" "$vlist"
           return 1
@@ -4417,7 +4487,7 @@
 
     if ! __trigger_validation "$uri" "$keyauthorization" "$vtype"; then
       _err "$d:Can not get challenge: $response"
-      _clearupwebbroot "$_currentRoot" "$removelevel" "$token"
+      _clearupwebbroot "$_currentRoot" "$removelevel" "$token" "$d"
       _clearup
       _on_issue_err "$_post_hook" "$vlist"
       return 1
@@ -4428,7 +4498,7 @@
         _debug "trigger validation code: $code"
       else
         _err "$d:Challenge error: $response"
-        _clearupwebbroot "$_currentRoot" "$removelevel" "$token"
+        _clearupwebbroot "$_currentRoot" "$removelevel" "$token" "$d"
         _clearup
         _on_issue_err "$_post_hook" "$vlist"
         return 1
@@ -4444,7 +4514,7 @@
       waittimes=$(_math "$waittimes" + 1)
       if [ "$waittimes" -ge "$MAX_RETRY_TIMES" ]; then
         _err "$d:Timeout"
-        _clearupwebbroot "$_currentRoot" "$removelevel" "$token"
+        _clearupwebbroot "$_currentRoot" "$removelevel" "$token" "$d"
         _clearup
         _on_issue_err "$_post_hook" "$vlist"
         return 1
@@ -4460,7 +4530,7 @@
       fi
       if [ "$?" != "0" ]; then
         _err "$d:Verify error:$response"
-        _clearupwebbroot "$_currentRoot" "$removelevel" "$token"
+        _clearupwebbroot "$_currentRoot" "$removelevel" "$token" "$d"
         _clearup
         _on_issue_err "$_post_hook" "$vlist"
         return 1
@@ -4475,7 +4545,7 @@
         _info "$(__green Success)"
         _stopserver "$serverproc"
         serverproc=""
-        _clearupwebbroot "$_currentRoot" "$removelevel" "$token"
+        _clearupwebbroot "$_currentRoot" "$removelevel" "$token" "$d"
         break
       fi
 
@@ -4495,7 +4565,7 @@
             _get "http://$d/.well-known/acme-challenge/$token" "" 1
           fi
         fi
-        _clearupwebbroot "$_currentRoot" "$removelevel" "$token"
+        _clearupwebbroot "$_currentRoot" "$removelevel" "$token" "$d"
         _clearup
         _on_issue_err "$_post_hook" "$vlist"
         return 1
@@ -4507,7 +4577,7 @@
         _info "Processing"
       else
         _err "$d:Verify error:$response"
-        _clearupwebbroot "$_currentRoot" "$removelevel" "$token"
+        _clearupwebbroot "$_currentRoot" "$removelevel" "$token" "$d"
         _clearup
         _on_issue_err "$_post_hook" "$vlist"
         return 1

